### ISSUE TITLE
Additional disk device (/dev/sdb) on Multipass instance

### ISSUE THREAD

**Issue by javisnche** [canonical/multipass#2303]
How do I add an additional disk device on Multipass instance? 
Multipass already comes with /dev/sda1 which has its own partitions.
What I am looking is to create another device disk like /dev/sdb and /dev/sdc.

**Comment by Saviq** [comment]
@javisnche why do you want those extra disks? They would just end up as extra files on the host file system.

You can use `launch --disk` to choose the size of the default disk, if you're lacking space.

I would like to understand what can't you achieve with just one disk.

**Comment by javisnche** [comment]
If the extra disk capability will be implemented, then that will be a big plus. It is a requirement from my job side of things to have at least a backup of my environment. And I can only back up on an additional disk. 
A few days ago, my multipass environment was corrupted and I lost all my data. This happened because of lack of back up.

**Comment by cryptosven76** [comment]
this feature would be great as in many cases you want more than 1 disk attached to a VM

**Comment by javisnche** [comment]
> What I think we could do is accept multiple values of `--disk`, the first one would still refer to the main drive, the extra ones could create additional disks.
> 
> It's not something near the top of our roadmap, though, so contributions welcome.
> 
> We're working on instance modification now, allowing to _modify_ the size of the main disk, but I don't think we'd be implementing that for any extra disks.



> What I think we could do is accept multiple values of `--disk`, the first one would still refer to the main drive, the extra ones could create additional disks.
> 
> It's not something near the top of our roadmap, though, so contributions welcome.
> 
> We're working on instance modification now, allowing to _modify_ the size of the main disk, but I don't think we'd be implementing that for any extra disks.
This scenario did not work

**Comment by javisnche** [comment]
If I launch a multipass instance say with 50G disk space, 

multipass launch --cpus 2 --disk 50G --mem 6G --name secondary

Does this take away 50G immediately from your Mac's volume available storage space?

Does it also immediately take away the cpus (s) and/or the mem.

**Comment by Saviq** [comment]
@javisnche no, disk is sparse - it will grow up (and slightly beyond, until it cleans up) to the requested size. CPUs are also dynamic (you don't "give" CPUs to instances, but just run "up to" that many tasks on your host's CPU). Memory will also only be allocated as the instance needs it (although that differs between hypervisors), but other than through restarting the instance it won't get freed up again.

**Comment by javisnche** [comment]
> @javisnche no, disk is sparse - it will grow up (and slightly beyond, until it cleans up) to the requested size. CPUs are also dynamic (you don't "give" CPUs to instances, but just run "up to" that many tasks on your host's CPU). Memory will also only be allocated as the instance needs it (although that differs between hypervisors), but other than through restarting the instance it won't get freed up again.

Thanks Saviq

**Comment by javisnche** [comment]
I noticed that Multipass instances location on Mac is /var/root/Library/Application Support/multipassd/vault/instances
Can you please clarify my doubts with the following questions?
a. Is there a way to not run multipass as root?
c. There is some image there in the above path - How can I locate this image, and and is it possible to open this image on Virtual Box?
d. If I am able to open the image on Virtual Box, does it also come with the data lets say data within /home/ubuntu. Say I write  files on my multipass /home/ubuntu/a.txt - does this a.txt also comes when I import the image in VirtualBox?

**Comment by Saviq** [comment]
> a. Is there a way to not run multipass as root?

Not at the moment, no.

>  c. There is some image there in the above path - How can I locate this image, and and is it possible to open this image on Virtual Box?

See https://multipass.run/docs/using-virtualbox-in-multipass-macos

> d. If I am able to open the image on Virtual Box, does it also come with the data lets say data within /home/ubuntu. Say I write files on my multipass /home/ubuntu/a.txt - does this a.txt also comes when I import the image in VirtualBox?

You don't _import_ it into VirtualBox, it's there in the first place, if you use the virtualbox driver.

**Comment by javisnche** [comment]
> > a. Is there a way to not run multipass as root?
> 
> Not at the moment, no.
> 
> > c. There is some image there in the above path - How can I locate this image, and and is it possible to open this image on Virtual Box?
> 
> See https://multipass.run/docs/using-virtualbox-in-multipass-macos
> 
> > d. If I am able to open the image on Virtual Box, does it also come with the data lets say data within /home/ubuntu. Say I write files on my multipass /home/ubuntu/a.txt - does this a.txt also comes when I import the image in VirtualBox?
> 
> You don't _import_ it into VirtualBox, it's there in the first place, if you use the virtualbox driver.

Thanks @Saviq

**Comment by cryptosven76** [comment]
i am also looking for this feature for development purposes, would be great to add that feature !

**Comment by spTorin** [comment]
As an example, if you want to test the operation of a MinIO distributed cluster, then it is impossible to do with a single root device.
"Error: Drive `http://minio-1.multipass:9091/data_2` is part of root drive, will not be used"

Please add support for multiple drives.

**Comment by lucj** [comment]
Is there any update on this one ?

**Comment by ksully-execute** [comment]
any update on this issue? It's important to many to have the ability to define multiple hard disks (raw and attached)....

**Comment by cryptosven76** [comment]
@Saviq would. you be willing to provide guidance for somebody to add this feature ?

**Comment by ricab** [comment]
Hi @cryptosven76, I am afraid this would require a good deal of time from our side, to provide design and implementations for the different backends and platforms. I agree it would be nice to have this, but our hands are pretty full of other nice things at the moment. We'll be sure to discuss this internally for upcoming roadmap plans though.

I am removing the "help wanted" tag because I think that is misleading in this case.

**Comment by zarinfam** [comment]
You can add a new disk using VirtualBox GUI app to your instance (`sudo VirtualBox`). I did it on mac os with VirtualBox driver, and it worked after starting the instance I had `/dev/sdb`.

**Comment by cryptosven76** [comment]
@Saviq @ricab any news on this requested feature. you mentioned in Nov there are no cycles but it will be added to the roadmap. anything you can share on this ?

**Comment by ricab** [comment]
Hi @cryptosven76, no news on this yet, I am afraid. We did discuss it when planning the last roadmap cycle, but we did not have space to include it quite yet.

**Comment by lucj** [comment]
Out of curiosity, could a cloud-init file be used to specify the disks / partitions of a Multipass VM ?

**Comment by townsend2010** [comment]
Hey all,

We plan in 6 month cycles that roughly align with the Ubuntu releases.  This did not make our current cycle which ends in October.  We will consider it again when planning for the next cycle.  Part of this is the actual user experience which are very deliberate about.  We strive for a simple interface and as such, we need to think about the best way to make this happen without overcomplicating the CLI.  Also, we have to consider multiple platforms with multiple hypervisors and make this work in such a way that it is transparent to the user what platform and hypervisor they are using.

Regarding cloud-init, that would only help partition and initialize a filesystem on a disk that is already attached to the virtual machine.  The hypervisor has to be told to pass in the extra disk when booting the OS.

**Comment by ChaoShi036** [comment]
> I have a similar requirement. I want to do some development and want to setup the disk as ceph osd disks.

Hi there, have you solve the problem and deploy the ceph successfully?

**Comment by tgdfool2** [comment]
I switched to `minikube` :-)

**Comment by ChaoShi036** [comment]
> I switched to `minikube` :-)

cool, thank you for response.

**Comment by mrmercc** [comment]
hi, is there any update on this?

**Comment by ricab** [comment]
HI @mrmercc, I am afraid not. We would very much like to do this, but we are a small team and we just haven't been able to allocate the resources for it. Hopefully in a not too distant future.

**Comment by uwydoc** [comment]
a solid usecase would be like this: create a vhd(x) file on another disk on the host machine, possibly ssd, and mount it into the vm as the root dir for development

**Comment by wundersmiths** [comment]
I needed this to test our microceph on Windows - you can use hyper-v to add an additional disk after the vm is created.

**Comment by jimporter** [comment]
Based on the above discussions, I think this is still a valid enhancement, but not one that we'd put on our roadmap just yet.
